lib/src/controls/surfaces/commandbar.dart

```diff
@@ -0,0 +1,552 @@
+      w = Row(children: [Expanded(child: w)]);
+    }
+    return w;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.compactBreakpointWidth == null) {
+      return _buildForPrimaryMode(context, CommandBarItemDisplayMode.inPrimary);
+    } else {
+      return LayoutBuilder(
+        builder: (context, constraints) {
+          if (constraints.maxWidth > widget.compactBreakpointWidth!) {
+            return _buildForPrimaryMode(
+                context, CommandBarItemDisplayMode.inPrimary);
+          } else {
+            return _buildForPrimaryMode(
+                context, CommandBarItemDisplayMode.inPrimaryCompact);
+          }
+        },
+      );
+    }
+  }
+}
+
+/// When a [CommandBarItem] is being built, indicates the visual context
+/// in which the item is being built.
+enum CommandBarItemDisplayMode {
+  /// The item is displayed in the horizontal area (primary command area)
+  /// of the command bar.
+  ///
+  /// The item should be rendered by wrapping content in a
+  /// [CommandBarItemInPrimary] widget.
+  inPrimary,
+
+  /// The item is displayed in the horizontal area (primary command area)
+  /// of the command bar, but it is requested that the item take up less
+  /// horizontal space so that more items may fit without overflow.
+  ///
+  /// The item should be rendered by wrapping content in a
+  /// [CommandBarItemInPrimary] widget.
+  inPrimaryCompact,
+
+  /// The item is displayed within the secondary command area (within a
+  /// Flyout as a drop down of the "more" button).
+  ///
+  /// Normally you would want to render an item in this visual context as a
+  /// [TappableListTile].
+  inSecondary,
+}
+
+/// Signature of function that can customize the widget returned by
+/// a CommandBarItem built in the given display mode. Can be useful to
+/// wrap the widget in a [Tooltip] etc.
+typedef CommandBarItemWidgetBuilder = Widget Function(
+    BuildContext context, CommandBarItemDisplayMode displayMode, Widget w);
+
+/// An individual control displayed within a [CommandBar]. This widget ensures
+/// that the child widget has the proper margin so the item has the proper
+/// minimum height and width expected of a control within a [CommandBar].
+abstract class CommandBarItem with Diagnosticable {
+  final Key? key;
+
+  /// Specify to allow customization of the built widget for a given
+  /// display mode. For example, to wrap a widget in a [Tooltip].
+  final CommandBarItemWidgetBuilder? widgetBuilder;
+  const CommandBarItem({
+    required this.key,
+    required this.widgetBuilder,
+  });
+
+  /// Builds the final widget for this display mode for this item.
+  /// This applies the widgetBuilder callback to the internally built
+  /// widget before returning.
+  Widget build(BuildContext context, CommandBarItemDisplayMode displayMode) {
+    var w = buildWidget(context, displayMode);
+    if (widgetBuilder != null) {
+      w = widgetBuilder!(context, displayMode, w);
+    }
```

Comment:
Currently, `buildWidget` will always be called. This could cause performance issues and memory leak. With the suggestion, `buildWidget` will only be called if necessary
```suggestion
    if (widgetBuilder != null) {
      return widgetBuilder!(context, displayMode, w);
    } else {
      return buildWidget(context, displayMode);
    }
```

Reply:
Now that this logic has been moved into a separate type of item, `CommandBarBuilderItem`, this logic will only be used when this type of item is explicitly used, which requires a non-null callback, so it should solve for this as well.